pkgs/development/tools/godot/default.nix

```diff
@@ -1,108 +1,136 @@
-{ stdenv
-, lib
+{ alsa-lib
+, alsa-plugins
 , fetchFromGitHub
-, scons
-, pkg-config
-, udev
+, freetype
+, installShellFiles
+, lib
+, libGLU
+, libpulseaudio
 , libX11
 , libXcursor
+, libXext
+, libXfixes
+, libXi
 , libXinerama
 , libXrandr
 , libXrender
-, libpulseaudio
-, libXi
-, libXext
-, libXfixes
-, freetype
-, openssl
-, alsa-lib
-, alsa-plugins
 , makeWrapper
-, libGLU
-, zlib
+, openssl
+, pkg-config
+, scons
+, stdenv
+, udev
 , yasm
-, withUdev ? true
+, zlib
 }:
 
-let
-  options = {
-    touch = libXi != null;
-    pulseaudio = false;
-    udev = withUdev;
-  };
-in
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (self: {
   pname = "godot";
   version = "3.5.1";
 
   src = fetchFromGitHub {
     owner = "godotengine";
     repo = "godot";
-    rev = "${version}-stable";
+    rev = "${self.version}-stable";
     sha256 = "sha256-uHwTthyhfeQN0R1XjqZ+kGRa5WcpeQzA/DO9hZk4lvU=";
   };
 
-  nativeBuildInputs = [ pkg-config makeWrapper ];
-  buildInputs = [
+  nativeBuildInputs = [
+    installShellFiles
+    pkg-config
+    makeWrapper
     scons
-    udev
+  ];
+
+  buildInputs = [
+    alsa-lib
+    freetype
+    libGLU
+    libpulseaudio
     libX11
     libXcursor
+    libXext
+    libXfixes
+    libXi
     libXinerama
     libXrandr
     libXrender
-    libXi
-    libXext
-    libXfixes
-    freetype
     openssl
-    alsa-lib
-    libpulseaudio
-    libGLU
-    zlib
+    udev
     yasm
+    zlib
   ];
 
-  patches = [ ./pkg_config_additions.patch ./dont_clobber_environment.patch ];
```

Comment:
Why is the no-clobbering environment patch not needed anymore? The code is still in-tree.

Reply:
In the case of `pkg_config_additions.patch`, I removed that because the file it was patching was different enough that the patch was irrelevant.  But in the case of `dont_clobber_environment.patch`, it was just because it was unclear what purpose this was serving.  I removed it with the intent of restoring it with additional explanatory comments if testing showed the need for it, but everything seemed to work fine without it.

I do see the original commit that added it indicates it was to "stop build process from looking in global directories", but I'm not sure what problem prompted that, because it builds just fine...

I'm happy to re-add the patch file, I would just want to have an understanding of what it's for before doing so.

Reply:
> I'm happy to re-add the patch file, I would just want to have an understanding of what it's for before doing so.

Yeah, same -- that's why I mentioned it. I want to make sure we're not breaking something by removing it, though maybe it's just not required anymore for whatever reason.

Reply:
I'm just leaving this removed for now.  If in the future it is proven to be needed, we can re-add it with explanatory comments.

Reply:
Hey, long time no see, hope all has been well.

I looked into this, and it looks like it is still required.

The lines that were patched out here were introduced in https://github.com/godotengine/godot/commit/b7ebd22314f72c1021bd75efd14583089027558f.

Essentially, a default environment is populated based on the OS, which includes global paths: https://github.com/SCons/scons/blob/79dcabc57575bc9001df32b31635ca8e4bf0fbf8/SCons/Platform/posix.py#L85-L115.

So, it makes sense to override them with the values from our environment, instead of appending to them as done in the original SCons file. In fact, I'd argue that we should probably override the environment entirely/get rid of it?

Let me know if that makes sense.

Reply:
Ah, just saw this one...  The question here is: What does this buy us if it builds without the patch?  I'd like to be able to put an explanation alongside the patch, but the fact that it builds regardless suggests that the environment it uses is just fine as is...

Reply:
@Twey I see you were involved in this "dont_clobber_environment.patch" patch a few years ago.  winterqt found this comment from you:  https://github.com/NixOS/nixpkgs/pull/66349#issuecomment-544138856.  But we're a bit confused about this.  For one, winterqt found no evidence that SCons ever had PKG_CONFIG_PATH in its default environment.  And second, I would like an explanation of this patch in terms of what the practical benefits are.  For instance, "in scenario X, the build fails without this patch".  I was able to build without the patch, so I'm inclined to remove it, but I wanted to see if you remembered some scenario that was broken without the patch.  It might be that it used to be needed but isn't anymore.

Reply:
cc-wrapper [does enforce purity](https://github.com/NixOS/nixpkgs/blob/1f9748835ce2b81a8ee14e5caa7c0cca062d9619/pkgs/build-support/cc-wrapper/cc-wrapper.sh#L69-L96) with regards to include paths and such... so we should be fine dropping it?

Reply:
The way that this patch works is a little confusing, I probably should have written more comments about it.
The original code creates a new `Environment` and then appends keys to it from the OS build environment (i.e. the nixpkgs environment).  However, the default SCons `Environment` (here `env_base`) includes a bunch of well-known paths like `/usr/bin` et cetera, and therefore the environment constructed by the original script will check those paths first, even if we try to override them in nixpkgs.  This would cause breakage when the binaries the build expects are found in `/usr`, for example when building in a `nix-shell` or a Nix build with `sandbox = false` under a non-NixOS distro.

The purpose of the patch is to change that code, amusingly clobbering the defaults set in `Environment` so that they don't clobber the paths we're trying to pass in from nixpkgs.

Back in the day, the gcc wrapper would respond to finding a `badPath` in linking by throwing an error and completely refusing to perform the offending link, which obviously broke the build in the situation described above.  It looks like it's more forgiving now, just skipping `badPath`s instead of halting the whole build, but removing this patch could still lead to some surprising behaviour in edge cases where paths make it into the build from the host system that are not just related to linking, as could happen by reading the wrong `.pc` file from `PKG_CONFIG_PATH` (which, as I recall, is one of the default paths added to new `Environment`s by SCons).